src/runtime-rs/crates/hypervisor/src/qemu/cmdline_generator.rs

```diff
@@ -1700,6 +1701,182 @@ impl ToQemuParams for ObjectIoThread {
     }
 }
 
+// PCIeRootPortDevice directly attached onto the root bus
+// -device pcie-root-port,id=rp0,bus=pcie.0,chassis=0,slot=0,multifunction=off,pref64-reserve=<X>B,mem-reserve=<Y>B
+#[derive(Debug, Default)]
+pub struct PCIeRootPortDevice {
+    id: String,
+    bus: String,
+    chassis: String,
+    slot: String,
+    multifunction: bool,
+    addr: String,
+
+    mem_reserve: String,
+    pref64_reserve: String,
+}
+
+impl PCIeRootPortDevice {
+    #[allow(clippy::too_many_arguments)]
+    fn new(
+        id: &str,
+        bus: &str,
+        chassis: &str,
+        slot: &str,
+        multifunc: bool,
+        addr: &str,
+        mem_reserve: u64,
+        pref64_reserve: u64,
+    ) -> Self {
+        PCIeRootPortDevice {
+            id: id.to_string(),
+            bus: bus.to_string(),
```

Comment:
To comply with the conventional constructor postcondition and guarantee that on constructor exit, the instance is fully initialised, I think we'd better do something like
```
bus: if bus.is_empty() { DEFAULT_PCIE_ROOT_BUS.to_owned() } else { bus.to_string() },
```
here.  The same goes for the rest of the fields that represent mandatory parameters (i.e. all but `mem_reserve` and `pref64_reserve`).  This would also make `qemu_params()` implementation more straightfoward, just serialising members to the command line unconditionally.